curriculum/challenges/english/25-front-end-development/review-react-forms-data-fetching-and-routing/67c9e932c6c4234532d46394.md

```diff
@@ -0,0 +1,426 @@
+      <li key={post.id}>{post.title}</li>
+    ))}
+  </ul>
+);
+```
+
+If you want to use Axios, you need to install and import it:
+
+```bash
+npm i axios
+```
+
+```js
+import axios from "axios";
+```
+
+Then you can fetch the data using `axios.get`:
+
+```js
+const [data, setData] = useState(null);
+const [loading, setLoading] = useState(true);
+const [error, setError] = useState(null);
+
+useEffect(() => {
+  const fetchData = async () => {
+    try {
+      const res = await axios.get(
+        "https://jsonplaceholder.typicode.com/users"
+      );
+      setData(res.data);
+    } catch (err) {
+      setError(err);
+    } finally {
+      setLoading(false);
+    }
+  };
+
+  fetchData();
+}, []);
+```
+
+To fetch data using the `useSWR` hook, you need to first install and import it.
+
+```bash
+npm i swr
+```
+
+```js
+import useSWR from "swr";
+```
+
+Here is how you can use the hook to fetch data:
+
+```jsx
+import useSWR from "swr";
+
+const fetcher = (url) => fetch(url).then((res) => res.json());
+
+const FetchTodos = () => {
+ const { data, error } = useSWR(
+   "https://jsonplaceholder.typicode.com/todos",
+   fetcher
+ );
+
+ if (!data) {
+   return <h2>Loading...</h2>;
+ }
+ if (error) {
+   return <h2>Error: {error.message}</h2>;
+ }
+
+ return (
+   <>
+     <h2>Todos</h2>
+     <div>
+       {data.map((todo) => (
+         <h3 key={todo.id}>{todo.title}</h3>
+       ))}
+     </div>
+   </>
+ );
+};
+
+export default FetchTodos;
+```
+
+## Working with the `useMemo` Hook
+
+- **Memoization**: This is an optimization technique in which the result of expensive function calls are cached (remembered) based on specific arguments. When the same arguments are provided again, the cached result is returned instead of re-computing the function.
+- **`useMemo` Hook**: This hook is used to memoize computed values. Here is an example of memoizing the result of sorting a large array. The `expensiveSortFunction` will only run when `largeArray` changes:
+
+```js
+const memoizedSortedArray = useMemo(
+  () => expensiveSortFunction(largeArray),
+  [largeArray]
+);
+```
+
+## Working with the `useCallback` Hook
+
+- **`useCallback` Hook**: This is used to memoize function references.
+
+```js
+const handleClick = useCallback(() => {
+ // code goes here
+}, [dependency]);
+``` 
+
+- **`React.memo`**: This is used to memoize a component to prevent it from unnecessary re-renders when its prop has not changed.
+
+```js
+const MemoizedComponent = React.memo(({ prop }) => {
+ return (
+   <>
+     {/* Presentation */}
+   </>
+ )
+});
+``` 
+
+## Dependency Management Tools
+
+- **Dependency Definition**: In software, a dependency is where one component or module in an application relies on another to function properly. Dependencies are common in software applications because it allows developers to use pre-built functions or tools created by others. The two core dependencies needed for a React project will be the `react` and `react-dom` packages:
```

Comment:
```suggestion
- **Dependency Definition**: In software, a dependency is where one component or module in an application relies on another to function properly. Dependencies are common in software applications because they allow developers to use pre-built functions or tools created by others. The two core dependencies needed for a React project will be the `react` and `react-dom` packages:
```